src/dmd/root/longdouble.d

```diff
@@ -81,10 +81,75 @@ void ld_clearfpu()
     }
 }
 
+//////////////////////////////////////////////////////////////
+
+unittest // impure due to accessing mutable globals (ld_pi etc.)
+{
+    import core.stdc.string;
+    import core.stdc.stdio;
+
+    char[32] buffer;
+    ld_sprint(buffer.ptr, 'a', ld_pi);
+    assert(strcmp(buffer.ptr, "0x1.921fb54442d1846ap+1") == 0);
+
+    ld_sprint(buffer.ptr, 'g', longdouble_soft(2.0));
+    assert(strcmp(buffer.ptr, "2.00000") == 0);
+
+    ld_sprint(buffer.ptr, 'g', longdouble_soft(1234567.89));
+    assert(strcmp(buffer.ptr, "1.23457e+06") == 0);
+
+    longdouble_soft ldb = longdouble_soft(0.4);
+    long b = cast(long)ldb;
+    assert(b == 0);
+
+    b = cast(long)longdouble_soft(0.9);
+    assert(b == 0);
+
+    long x = 0x12345678abcdef78L;
+    longdouble_soft ldx = longdouble_soft(x);
+    assert(ldx > ld_zero);
+    long y = cast(long)ldx;
+    assert(x == y);
+
+    x = -0x12345678abcdef78L;
+    ldx = longdouble_soft(x);
+    assert(ldx < ld_zero);
+    y = cast(long)ldx;
+    assert(x == y);
+
+    ulong u = 0x12345678abcdef78L;
+    longdouble_soft ldu = longdouble_soft(u);
+    assert(ldu > ld_zero);
+    ulong v = cast(ulong)ldu;
+    assert(u == v);
+
+    u = 0xf234567812345678UL;
+    ldu = longdouble_soft(u);
+    assert(ldu > ld_zero);
+    v = cast(ulong)ldu;
+    assert(u == v);
+
+    u = 0xf2345678;
+    ldu = longdouble_soft(u);
+    ldu = ldu * ldu;
+    ldu = sqrt(ldu);
+    v = cast(ulong)ldu;
+    assert(u == v);
+
+    u = 0x123456789A;
+    ldu = longdouble_soft(u);
+    ldu = ldu * longdouble_soft(1L << 23);
+    v = cast(ulong)ldu;
+    u = u * (1L << 23);
+    assert(u == v);
+}
+
+//////////////////////////////////////////////////////////////
+
 pure:
-@safe:
+@trusted: // LDC: LLVM __asm is @system AND requires taking the address of variables
```

Comment:
`@trusted:` at the top of a module is just asking for trouble

Reply:
Not to happy about it, but current state is also stetching the concept of safety a bit, because separate trusted asm sections are combined in each function. So maybe it is ok to mark most of the module as `@trusted`.

You can reenable `@safe` before the constants starting with `ld_qnan`, though.

Reply:
Done.